### ISSUE TITLE
TST: Migrate to CircleCI for builds

### ISSUE THREAD

**Issue by tswast** [pydata/pandas-gbq#228]
Everytime I use Travis, I have the frustration that I can't really define separate Docker images to use as the base for conda versus pip versus different versions of Python. This made changes like https://github.com/pydata/pandas-gbq/pull/222 a lot harder than they would have been if I could have had more control over the base image.

I also suspect that Travis is maybe at fault for our Conda build failing https://github.com/pydata/pandas-gbq/issues/189. My only evidence for this is that I've tested on conda locally several times and am unable to reproduce the issue we see in the Travis build.

There are two interacting features of CircleCI that I think would be a big help:

1. [Docker executor type](https://circleci.com/docs/2.0/executor-types/#using-docker)
2. [Workflows](https://circleci.com/docs/2.0/workflows/), which provide the ability to run multiple "jobs" in parallel, where each job has its own executor definition.

One problem with this plan is that I don't have access to the GCP project that we use for Travis builds so that I can set up the secrets on CircleCI. Maybe @jreback could give me access to that?

**Comment by tswast** [comment]
pandas is mostly migrating to Azure Pipelines: https://github.com/pandas-dev/pandas/issues/22690

**Comment by max-sixty** [comment]
Azure would allow us to test Windows too.

We use Google Cloud Builder successfully internally - but I'm not sure there's a mature OSS offering. Is that right?

**Comment by tswast** [comment]
> Azure would allow us to test Windows too.

Yeah. I'm not too keen on having to maintain VM images, but it'll be okay if we can re-use the configs from pandas.

> We use Google Cloud Builder successfully internally - but I'm not sure there's a mature OSS offering. Is that right?

There's a free tier for Cloud Builder, but nothing OSS-specific AFAIK.

**Comment by tswast** [comment]
I just mailed https://github.com/pydata/pandas-gbq/pull/232. The reason I went with CircleCI instead of Azure is that I couldn't "install" the Azure GitHub app. Even though it's free for OSS projects, I can't install it due to not being a billing admin for the pydata org. CircleCI does not have this requirement. I can enable builds directly from the CircleCI interface without going through GitHub for "billing".

Also, I'm quite pleased by the [local development capabilities of CircleCI](https://circleci.com/docs/2.0/local-cli/). The fact that everything is run in Docker containers makes it very easy to test locally with an almost-identical environment before pushing my changes to GitHub. I don't think we'll run into the problems that pandas had with CircleCI because we aren't doing any complicated native-code compilation.
